pennylane/devices/preprocess.py

```diff
@@ -545,3 +544,211 @@ def _validate_snapshot_shots(tape, sample_measurements, analytic_measurements, n
+        samples = results[0]
+        results_processed = []
+        for m in tape.measurements:
+            if tape.shots.has_partitioned_shots:
+                res = tuple(m.process_samples(unsqueezed(s), measured_wires) for s in samples)
+            else:
+                res = m.process_samples(unsqueezed(samples), measured_wires)
+            results_processed.append(res)
+
+        if len(tape.measurements) == 1:
+            results_processed = results_processed[0]
+        else:
+            results_processed = tuple(results_processed)
+        return results_processed
+
+    return [new_tape], postprocessing_fn
+
+
+@transform
+def measurements_from_counts(tape):
+    r"""Quantum function transform that replaces all measurements from a tape with a single
+    counts measurement, and adds postprocessing functions for each original measurement.
+
+    This transform can be used to make tapes compatible with device backends that only return
+    `counts`. The final output will return the initial requested measurements, calculated from
+    the raw counts returned immediately after execution.
+
+    The transform is only applied if the tape is being executed with shots.
+
+    .. note::
+        This transform diagonalizes all the operations on the tape. An error will
+        be raised if non-commuting terms are encountered. To avoid non-commuting
+        terms in circuit measurements, the :func:`split_non_commuting <pennylane.transforms.split_non_commuting>`
+        transform can be applied.
+
+    Args:
+        tape (QNode or QuantumTape or Callable): A quantum circuit.
+
+    Returns:
+        qnode (QNode) or quantum function (Callable) or tuple[List[QuantumTape], function]: The
+        transformed circuit as described in :func:`qml.transform <pennylane.transform>`.
+
+    **Example**
+
+    Consider the tape:
+
+    >>> ops = [qml.X(0), qml.RY(1.23, 1)]
+    >>> measurements = [qml.expval(qml.Y(0)), qml.probs(wires=[1])]
+    >>> tape = qml.tape.QuantumScript(ops, measurements, shots=10)
+
+    We can apply the transform to diagonalize and convert the two measurements to a single sample:
+
+    >>> (new_tape, ), fn = qml.transforms.measurements_from_counts(tape)
+    >>> new_tape.measurements
+    [CountsMP(wires=[0, 1], all_outcomes=False)]
+
+    The tape operations now include diagonalizing gates.
+
+    >>> new_tape.operations
+    [X(0), RY(1.23, wires=[1]), RX(1.5707963267948966, wires=[0])]
+
+    The tape is now compatible with a device backend that only supports counts. Executing the
+    tape returns the raw counts:
+
+    >>> dev = qml.device("default.qubit")
+    >>> dev.execute(new_tape)
+    {'00': 4, '01': 2, '10': 2, '11': 2}
+
+    And these can be post-processed to get the originally requested measurements:
+
+    >>> fn((res,))
+    (-0.19999999999999996, array([0.7, 0.3]))
+    """
+    if tape.shots.total_shots is None:
+        print("returning the tape, because shots is None")
```

Comment:
```suggestion
        warnings.warn("returning the tape, because shots is None", UserWarning)
```
I guess print will be suppressed runtime?

Reply:
Wait do users need to read this info?

Reply:
Hahaha, no, this is from me debugging a silly mistake yesterday. I will delete it.